python_modules/libraries/dagster-dbt/dagster_dbt_tests/core/dbt_packages/test_columns_metadata.py

```diff
@@ -579,6 +589,17 @@ def my_dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource):
         str(column_lineage_by_asset_key) + "\n\n" + str(expected_column_lineage_by_asset_key)
     )
 
+    # Ensure we cache relation metadata fetches
+    if use_experimental_fetch_column_schema:
+        relation_keys_passed = [
+            call.kwargs["relation_key"] for call in get_relation_from_adapter.call_args_list
+        ]
+        # We may query the same relation multiple times if they initiate at around the same time.
+        # Still, the total number of unique relations queried should be a lot lower than the total
+        # number of instances where we get column metadata (around 33 instead of 60+)
+        REPEAT_QUERIES_PADDING = 10
```

Comment:
If we do a three step pass, we won’t run into this problem.

- First pass: for each asset materialization, fetch the column schema metadata and cache
- Second pass: find the parents of the source nodes of the executed DAG. These parents weren’t fetched (either because they aren’t executed due to subselection, or they were modeled as dbt sources) so fetch them and cache.
- Third pass: for each asset materialization, calculate the column lineage using the cached column schema. The cache will always be hit.

The source nodes of executed DAG will never have their parents

Reply:
I think you're right in that there's a way to do this without double reads, but doing so does introduce more complexity and synchronization between threads. One of the nice factors of the implementation as-is is that each materialization can kick off its metadata fetch as soon as a thread in the pool is available, it doesn't need to concern itself about the status of any other fetch.

I think the approach above requires us to wait for the metadata fetch of each parent node to be finished before we can begin the fetch of downstream nodes, right? Can look into reworking the execution to be more deliberate to prevent double-reads down the line but I think that "mostly-cached" is a big step up from no cache.

Reply:
If we have assets `A`, `B`, `C`, `D`, with `C` downstream of `A` and `B`:

```

Time       A     B     C     D
 |              
 |         *     *           *
 |         *     *           *      
 |         *     *           *      
 |               *           *
 |         F                 *
 |         F           *     *  
 |         F     F     *     *  
 |               F           * 
 |               F     F     *  
 |                     F     *  
 |                     F     *  
 |                           *
 |
 V


* = Materializing
F = Fetching metadata
```

In this case the threadpool picks up the work item to fetch metadata for `C`. Without synchronization, it might double-read the metadata for its parent, B, if it's not cached by the time that the `C` metadata work item gets to processing its parents' data. I think to prevent it we'd need to make the `C` metadata work item aware of a dependency.

Reply:
> I think the approach above requires us to wait for the metadata fetch of each parent node to be finished before we can begin the fetch of downstream nodes, right?

No, I don't think so. On the first pass, the metadata fetch has no dependencies. We're just fetching the schema and caching.

The third pass will also have no dependencies since we'll entirely rely on the cache. So again, each metadata fetch will have no dependencies.

Reply:
Hmm, I see what you're saying, but having distinct "passes" will also require some form of synchronization, right? e.g. we need to wait for the first pass to finish before we can start the second pass and so on. Right now the deferred compute doesn't restrict operations from interleaving (e.g. we can be materializing some assets, fetching row counts for some, and fetching metadata for others).

Reply:
oops yeah you’re right, I forgot that this is essentially making all the materializations wait until the last one finishes. Disregard my comments!